### PR TITLE
TLS implementation 

### PR DESCRIPTION
**Pull request by MiaAltieri**
### For the reader 
Requesting to merge into `TLS-feature-branch` rather than into `main`, so that all of TLS (libs, implementation, unit-test, int-tests) can be merged at once and pushed to charmhub. TLS integration tests coming in next PR

### Problem

TLS not supported

### Solution

TLS now supported.

### Release Notes

1. Updated `.yaml` files & `requirements.txt`
2. `lib/charms/mongodb_libs/v0/helpers.py` updated to add/remove tls arguments for starting mongod -- does NOT need reviewed, this is part of the k8s lib requirments
3. `lib/charms/mongodb_libs/v0/machine_helpers.py` updated to add/remove tls arguments for starting mongod via systemd, and added additional machine charm helpers
4. `lib/charms/mongodb_libs/v0/mongodb_tls.py` updated to work with both k8s and VM
5. `src/charm.py` now adds TLS files if necessary

### Testing

Integration and Unit tests coming in following PR. 
This work was checked by hand by issuing this command on the associated mongodb unit/(s):
`mongo '<connection uri>' --eval 'rs.status()' --tls --tlsCAFile /etc/mongodb/external-ca.crt --tlsCertificateKeyFile /etc/mongodb/external-cert.pem`

### CODE REVIEW

File: lib/charms/mongodb_libs/v0/machine_helpers.py [hunk 1]
```diff
@@ -146,29 +154,64 @@ def generate_service_args(auth: bool, machine_ip: str, replset: str) -> str:
...
+            mongod_start_args.extend(
+                [
+                    f"--tlsCAFile={TLS_EXT_CA_FILE}",
+                    f"--tlsCertificateKeyFile={TLS_EXT_PEM_FILE}",
+                    # allow non-TLS connections
+                    "--tlsMode=preferTLS",
+                ]
+            )
+
+        # internal TLS can be enabled only in external is enabled
```
Comment (marceloneppel) [inline lib/charms/mongodb_libs/v0/machine_helpers.py:173]:
```suggestion
        # internal TLS can be enabled only if external is enabled
```

File: lib/charms/mongodb_libs/v0/mongodb_tls.py [hunk 2]
```diff
@@ -171,7 +182,17 @@ def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
             return
 
         logger.debug("Restarting mongod with TLS enabled.")
-        self.charm.on_mongod_pebble_ready(event)
+        if self.substrate == "vm":
+            self.charm._push_tls_certificate_to_workload()
+            self.charm.unit.status = MaintenanceStatus("disabling TLS")
```
Comment (marceloneppel) [inline lib/charms/mongodb_libs/v0/mongodb_tls.py:187]:
```suggestion
            self.charm.unit.status = MaintenanceStatus("enabling TLS")
```

File: lib/charms/mongodb_libs/v0/mongodb_tls.py [hunk 3]
```diff
@@ -129,7 +132,16 @@ def _on_tls_relation_broken(self, event: RelationBrokenEvent) -> None:
             return
 
         logger.debug("Restarting mongod with TLS disabled.")
-        self.charm.on_mongod_pebble_ready(event)
+        if self.substrate == "vm":
+            self.charm.unit.status = MaintenanceStatus("disabling TLS")
+            restart_mongod_service(
+                auth=True,
```
Comment (marceloneppel) [inline lib/charms/mongodb_libs/v0/mongodb_tls.py:138]: Just a question: is there a situation where auth should be disabled and this value needs to be `False`?

**Review by marceloneppel — APPROVED** [review]
Looks great, Mia!

File: lib/charms/mongodb_libs/v0/mongodb_tls.py (hunk 2 above)
Comment (MiaAltieri) [reply lib/charms/mongodb_libs/v0/mongodb_tls.py:187]: 🤦‍♀️ thanks for catching this!

File: lib/charms/mongodb_libs/v0/mongodb_tls.py (hunk 3 above)
Comment (MiaAltieri) [reply lib/charms/mongodb_libs/v0/mongodb_tls.py:138]: Great idea! I guess if someone is using the legacy relation and wants to use TLS, I've changed it now to use whatever the current `auth` is set to.

File: lib/charms/mongodb_libs/v0/machine_helpers.py [hunk 4]
```diff
@@ -146,29 +154,64 @@ def generate_service_args(auth: bool, machine_ip: str, replset: str) -> str:
...
+    with open(file_name, "w") as write_file:
+        write_file.write(file_contents)
+
+    os.chmod(file_name, 0o400)
+    mongodb_user = pwd.getpwnam(MONGO_USER)
+    os.chown(file_name, mongodb_user.pw_uid, mongodb_user.pw_gid)
+
+
+def restart_mongod_service(auth: bool, machine_ip: str, config: MongoDBConfiguration):
+    """Restarts the mongod service with its associated configuratiion."""
```
Comment (welpaolo) [inline lib/charms/mongodb_libs/v0/machine_helpers.py:210]: typo

**Review by welpaolo — APPROVED** [review]
Looks good :)

File: lib/charms/mongodb_libs/v0/machine_helpers.py [hunk 5]
```diff
@@ -146,29 +154,64 @@ def generate_service_args(auth: bool, machine_ip: str, replset: str) -> str:
...
 
+def push_file_to_unit(parent_dir, file_name, file_contents) -> None:
+    """K8s charms can push files to their containers easily, this is the vm charm workaround."""
+    Path(parent_dir).mkdir(parents=True, exist_ok=True)
+    with open(file_name, "w") as write_file:
+        write_file.write(file_contents)
+
+    os.chmod(file_name, 0o400)
+    mongodb_user = pwd.getpwnam(MONGO_USER)
+    os.chown(file_name, mongodb_user.pw_uid, mongodb_user.pw_gid)
```
Comment (paulomach) [inline lib/charms/mongodb_libs/v0/machine_helpers.py:206]: tip: you can use [`shutil.chown`](https://docs.python.org/3/library/shutil.html#shutil.chown) to use an one liner

File: lib/charms/mongodb_libs/v0/mongodb.py [hunk 6]
```diff
@@ -52,6 +52,8 @@ class MongoDBConfiguration:
     password: str
     hosts: Set[str]
     roles: Set[str]
+    tls_external: bool
```
Comment (paulomach) [inline lib/charms/mongodb_libs/v0/mongodb.py:55]: update docstring (L42-L48) to match all fields

**Review by paulomach — APPROVED** [review]
minor comments, looking good!

File: lib/charms/mongodb_libs/v0/mongodb.py (hunk 6 above)
Comment (MiaAltieri) [reply lib/charms/mongodb_libs/v0/mongodb.py:55]: good eye!

**Pull request merged**